gh-119180: PEP 649 compiler changes

JelleZijlstra opened the pull request:
This implements the compiler changes in PEP-649, creating a compiler-generated `__annotate__` function that holds the function, class, and module annotations.

I commented on Discuss on some of the CPython tests I needed to change: https://discuss.python.org/t/pep-649-deferred-evaluation-of-annotations-tentatively-accepted/21331/60

I tried this branch on a few prominent runtime typing projects; my notes are in the issue description (#119180). Mostly there are few issues, except a few places that need adjustment because they were checking for annotations in the `__dict__` directly.

<!-- gh-issue-number: gh-119180 -->
* Issue: gh-119180
<!-- /gh-issue-number -->

JelleZijlstra committed:
Some initial work

JelleZijlstra committed:
Add __annotate__ descriptors

JelleZijlstra committed:
fix refleaks

JelleZijlstra committed:
blurb

JelleZijlstra committed:
Fix some tests

JelleZijlstra committed:
regen globals

JelleZijlstra committed:
Add bytecode for adding annotate

JelleZijlstra committed:
compiler changes

JelleZijlstra committed:
Functions work

JelleZijlstra committed:
Raise AssertionError on the wrong format

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
Modules and classes

JelleZijlstra committed:
attempts

JelleZijlstra committed:
Fixes

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
Fix code object name

JelleZijlstra committed:
Start fixing test_type_annotations

JelleZijlstra committed:
Fix class scoping

JelleZijlstra committed:
unyielding

JelleZijlstra committed:
Fix test_typing

JelleZijlstra committed:
Fix test_type_parmas

JelleZijlstra committed:
Add test, add to inspect

JelleZijlstra committed:
Use inspect constants

JelleZijlstra committed:
test_grammar tweaks

JelleZijlstra committed:
fix test_positional_only_arg

JelleZijlstra committed:
Fix test_module

JelleZijlstra committed:
Fix symtable tests

JelleZijlstra committed:
fix test_pydoc

JelleZijlstra committed:
fix test_traceback

JelleZijlstra committed:
fix test_opcodes

JelleZijlstra committed:
Merge branch 'main' into pep649-compile

JelleZijlstra committed:
Raise NotImplementedError

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
blurb

JelleZijlstra committed:
Fix test_dis

JelleZijlstra committed:
Remove broken tests

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
No deferred evaluation in interactive mode

JelleZijlstra committed:
gh-119443: Turn off from __future__ import annotations in REPL

JelleZijlstra committed:
Fix refleak

JelleZijlstra committed:
Fix another refleak

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
Exit scope correctly on error

JelleZijlstra committed:
fix test

JelleZijlstra committed:
Initialize field

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
self-review

JelleZijlstra committed:
Fix crash found by CIFuzz

AlexWaygood commented on Lib/test/test_pydoc/test_pydoc.py:
```diff
@@ -76,6 +76,11 @@ class A(builtins.object)
+     |
+     |  __annotate__ = <annotations of B>(...)
+     |
+     |  ----------------------------------------------------------------------
```
This leads to some pretty confusing output if you call `help()` in the REPL on a class that has annotations in the class body:

![image](https://github.com/python/cpython/assets/66076021/9d8b1ddb-2834-4831-9c9c-383bbc777545)

I think `__annotate__ = (...)` would be fine, but `<annotations of B>(...)` looks like it's trying to be an invalid call expression of some kind

JelleZijlstra replied:
What do you suggest? Should we special case `__annotate__` in pydoc somewhere?

AlexWaygood replied:
This is fundamentally caused by the fact that `__annotate__` has its `__name__` set to a string with spaces in it, right? Why does `__annotate__` not have its `__name__` set to the name it can be accessed by (`__annotate__`), like nearly every other method that you can access from a class?

```pycon
>>> class Foo:
...     x: int
...     def y(self): pass
...     
>>> print(Foo.y, Foo.y.__name__, Foo.__annotate__, Foo.__annotate__.__name__, sep="\n")
<function Foo.y at 0x104e99430>
y
<function Foo.<annotations of Foo> at 0x104e992d0>
<annotations of Foo>
```

![image](https://github.com/python/cpython/assets/66076021/72fa587f-baa1-4976-a0e5-91effeb783fc)

JelleZijlstra replied:
I guess that means every `__annotate__` function would have the same name when it shows up in stack traces, which maybe isn't the worst?

It would mean that in a stack trace, instead of this:

```
% ./python.exe -c '''
def f(x: y): pass

print(f.__annotations__)
'''
Traceback (most recent call last):
  File "<string>", line 4, in <module>
    print(f.__annotations__)
          ^^^^^^^^^^^^^^^^^
  File "<string>", line 2, in <annotations of f>
    def f(x: y): pass
             ^
NameError: name 'y' is not defined
```

You'd see "`line 2, in __annotate__`".

AlexWaygood replied:
> I guess that means every `__annotate__` function would have the same name when it shows up in stack traces, which maybe isn't the worst?

I think that's probably not the worst. For most method calls, you only generally get the name of the method in the stack trace, you don't generally get the qualname:

```pycon
>>> class Foo:
...     def __init__(self):
...         causes_namerror
...         
>>> Foo()
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    Foo()
    ~~~^^
  File "<python-input-0>", line 3, in __init__
    causes_namerror
NameError: name 'causes_namerror' is not defined
```

I _like_ the `<annotations of f>` thing showing up in the stack trace -- since this is a function that has no source code, that has been magically inserted by the compiler, I do agree that it's helpful to display it like that. But IMHO, I think it's more important to maintain the invariant here that for a method in a class namespace, `__name__` is almost always a valid identifier, and that `__name__` is almost always the identifier by which you can access that method using the `.` operator on the class.

If we want to have tracebacks show `<annotations of f>` rather than `__annotate__`, maybe we could add some special-casing to `traceback.py`? But again, I don't think it's the worst thing in the world if the traceback just shows `__annotate__`.

carljm replied:
Alex's argument makes sense to me, and I agree with it as far as tracebacks go. The line number should be sufficient to clarify which `__annotate__`, and the frame we're talking about here is one that is just totally absent in the status quo, so we aren't making things worse.

But I think renaming to just `__annotate__` will lead to confusing symbol tables, at the very least, whenever multiple functions are defined in one scope. Since function `__annotate__` scopes are sibling to the function scope, if you have multiple functions defined in a scope you currently get this:

```
>>> code = """
... def f() -> None: ...
... def g() -> int: ...
... """
>>> st = symtable.symtable(code, "test", "exec")
>>> st
<SymbolTable for module test>
>>> st.get_children()
[<SymbolTable for <annotations of f> in test>, <Function SymbolTable for f in test>, <SymbolTable for <annotations of g> in test>, <Function SymbolTable for g in test>]
```

With the proposed change here, that last line would show two different `__annotate__` symbol tables, with no way (other than implicitly by ordering, I guess) to differentiate which one goes with which function.

I'm not sure in practice how much of a problem this would be.

JelleZijlstra replied:
I pushed a change renaming the function to `__annotate__`.

AlexWaygood commented on Lib/typing.py:
```diff
@@ -24,6 +24,7 @@
 import collections.abc
 import copyreg
 import functools
+import inspect
```
:/

But, if this is necessary, we should simplify the bits of `typing.py` elsewhere that are specifically dedicated to help us avoid importing `inspect` at the module level: https://github.com/python/cpython/blob/a531fd7fdb45d13825cb0c38d97fd38246cf9634/Lib/typing.py#L1974-L1979

JelleZijlstra replied:
I'd like to keep those changes out of this PR for now; there's a chance the Python-level code will change later so we don't need the `inspect` import.

AlexWaygood replied:
Okay. Is there an issue where we can track these things that we're deferring for now? Should we create a list in #119180? I worry that they'll get forgotten about otherwise.

JelleZijlstra replied:
Yes, I just added it to #119180; feel free to edit the list there with more things that need doing.

AlexWaygood commented on Lib/inspect.py:
```diff
@@ -173,6 +177,13 @@
 TPFLAGS_IS_ABSTRACT = 1 << 20
 
 
+@enum.global_enum
```
I know it's specified by PEP-649, but I'd honestly prefer it if this were not a global enum. To me, `inspect.AnnotationsFormat.VALUE` (or `AnnotationsFormat.VALUE`) is much clearer than `inspect.VALUE`. There are loads of things that have values; `inspect.VALUE` could refer to anything. And `inspect.py` already has a huge module-level API.

JelleZijlstra replied:
I don't disagree but for now I'm trying to implement the PEP as written. I think once the implementation is more complete we'll come up with more places where we may want to change the PEP's API.

carljm replied:
This would also be improved by putting the PEP 649 implementation (and `inspect.get_annotations`) into a new `annotools` (or whatever we call it) module.

AlexWaygood replied:
I've added a TODO as a sub-item in https://github.com/python/cpython/issues/119180#issue-2304664735 so we don't forget to revisit this later; marking this as resolved for now :-)

AlexWaygood commented on Lib/typing.py:
```diff
@@ -2970,7 +2971,12 @@ def __new__(cls, typename, bases, ns):
+        elif "__annotate__" in ns:
+            types = ns["__annotate__"](inspect.VALUE)
+        else:
+            types = {}
```
It feels unfortunate that this kind of metaprogramming becomes more complicated, but perhaps it's not a common enough use case for it to warrant its own helper function in `inspect` or `types`

JelleZijlstra replied:
Yes, I'll probably want to add a helper function for this somewhere. However, I'd like to focus this PR on the compiler changes (plus whatever else is needed to get tests passing); a future PR will focus on providing better Python-level abstractions.

JelleZijlstra replied:
Note that for full PEP 649 support this code is not sufficient, since it uses VALUE semantics to evaluate the entire annotations dict. That's something I plan to improve in a later PR.

iritkatriel commented on Python/compile.c:
```diff
@@ -1622,6 +1633,120 @@ compiler_unwind_fblock_stack(struct compiler *c, location *ploc,
+static int
+compiler_collect_annotations(struct compiler *c, asdl_stmt_seq *stmts,
+                             int *annotations_len)
+{
```
Do we really need to add a second pass to the compiler for this?

JelleZijlstra replied:
The alternative would I think be to set up some system where we are emitting instructions into two different code objects at the same time, which would likely be more complicated. Note that this second pass happens only on module and class bodies and does not recurse into AST nodes that cannot contain AnnAssign statements. I believe I also made it so the pass runs only if the symtable has detected that there are any annotations in this block.

iritkatriel replied:
I think there is another alternative - to collect the annotations into a list on the compiler unit during the normal pass, and then use this list to create the annotations code object. Would that not work?

I'm not concerned about performance, but about maintenance complexity. A second pass is just asking to get out of sync when we change the AST node set (that it's only considering some of the node types makes this more likely).

JelleZijlstra replied:
That's a reasonable idea. We in fact already have a pass like this in the `find_ann()` functions, and we've had some past bugs where annotations were dropped if they appeared inside some blocks because they were missed in `find_ann()`. I realized it's easy to get rid of this pass by recording some data in symtable.c; I'll push a commit for that.

However, the case I'm adding in this PR is a bit harder. AST nodes aren't PyObjects, so I can't put them in a list. I suppose I could convert AnnAssign pointers into Python ints and then stash them in a list.

carljm commented on Lib/symtable.py:
```diff
@@ -218,6 +218,8 @@ def get_methods(self):
         if self.__methods is None:
             d = {}
             for st in self._table.children:
+                if st.type == _symtable.TYPE_ANNOTATION:
```
This is really a PEP 695 follow-up, not part of 649, but I suspect we should also skip `TYPE_ALIAS` blocks here. Right now if you have `type T = int` in a class body, `get_methods()` will list `T` as one of the methods.

I'm guessing it was caught in 649 and not 695 because of the scope name in 649 not being an identifier.

JelleZijlstra replied:
This code is quite buggy; it also returns nested classes as "methods". That should perhaps be fixed but not in this PR.

picnixz replied:
Just wondering, but do you want perhaps to revert that change for now and let me include it in #120151 as a follow-up ?

JelleZijlstra replied:
No, it would leave test_symtable.py broken. We'll just have to deal with the merge conflict in whichever one gets merged second.

picnixz replied:
I can wait for you PR to be merged (I think it'd be cleaner and mine is not really pressing).

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
fix merge

JelleZijlstra committed:
Name the function as __annotate__

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

carljm commented on Lib/test/test_symtable.py:
```diff
@@ -205,12 +205,14 @@ def test_assigned(self):
     def test_annotated(self):
         st1 = symtable.symtable('def f():\n    x: int\n', 'test', 'exec')
-        st2 = st1.get_children()[0]
+        st2 = st1.get_children()[1]
```
It was a bit odd to me at first that for classes, their `__annotate__` symbol table shows up as a child of the class symbol table, but for functions, it shows up as a sibling of the function symbol table.

But I guess on further reflection this makes some sense. `__annotate__` is in fact part of the class namespace/scope. For functions, it is an attribute of the function object, but not actually part of the function scope, which is the namespace of the function body. So I guess a sibling scope is the most reasonable representation.

JelleZijlstra committed:
Replace find_ann()

JelleZijlstra committed:
fix test

JelleZijlstra committed:
Remove second compiler pass

JelleZijlstra committed:
Fix refleak

JelleZijlstra committed:
Fix a test

JelleZijlstra committed:
Fix bug when there are only non-simple annotations

carljm commented on Lib/test/test_type_annotations.py:
```diff
@@ -256,3 +265,120 @@ def check_annotations(self, f):
+                self.assertIsInstance(annotate, types.FunctionType)
+                self.assertEqual(annotate.__name__, "__annotate__")
+                with self.assertRaises(NotImplementedError):
+                    annotate(inspect.FORWARDREF)
```
any reason not to test `inspect.SOURCE` raises `NotImplementedError` here too?

carljm commented on Lib/typing.py:
```diff
@@ -3143,7 +3154,8 @@ def __new__(cls, name, bases, ns, total=True):
 
         for base in bases:
-            annotations.update(base.__dict__.get('__annotations__', {}))
+            # TODO: preserve laziness
```
I'm not sure what this TODO means exactly. Preserving laziness doesn't seem possible for `TypedDict`; it needs to look at the annotations. Using `inspect.FORWARDREFS` to allow forward references seems like it might make sense.

JelleZijlstra replied:
Pushing an expanded version of this comment

JelleZijlstra committed:
Fix more tests

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
Remove REPL special case

JelleZijlstra committed:
no docstrings in the repl

JelleZijlstra committed:
Fix pyrepl test

carljm commented on Python/symtable.c:
```diff
@@ -1342,21 +1355,20 @@ symtable_enter_block(struct symtable *st, identifier name, _Py_block_ty block,
 
     /* Annotation blocks shouldn't have any affect on the symbol table since in
      * the compilation stage, they will all be transformed to strings. They are
      * only created if future 'annotations' feature is activated. */
```
This comment needs updating.

carljm commented on Python/symtable.c:
```diff
@@ -2440,18 +2486,52 @@ symtable_visit_params(struct symtable *st, asdl_arg_seq *args)
 static int
-symtable_visit_annotation(struct symtable *st, expr_ty annotation)
+symtable_visit_annotation(struct symtable *st, expr_ty annotation,
+                          struct _symtable_entry *parent_ste, void *key)
```
Why do we need the `parent_ste` argument when the only caller always passes `st->st_cur`? I realize we need to preserve a pointer to `parent_ste` because we are going to push a block, but that can be done internal to the function.

carljm commented on Python/symtable.c:
```diff
@@ -2440,18 +2486,52 @@ symtable_visit_params(struct symtable *st, asdl_arg_seq *args)
+        }
+    }
+    else {
+        if (st->st_cur->ste_annotation_block == NULL) {
```
I had to check myself that `st->st_cur` here and `parent_ste` on line 2506 were the same object. Would it be clearer to use `parent_ste` here as well?

carljm commented on Python/symtable.c:
```diff
@@ -2467,37 +2547,58 @@ symtable_visit_argannotations(struct symtable *st, asdl_arg_seq *args)
-    if (future_annotations && !symtable_exit_block(st)) {
-        VISIT_QUIT(st, 0);
+    if (returns) {
+        st->st_cur->ste_annotations_used = 1;
+        VISIT(st, expr, returns);
     }
-    if (returns && !symtable_visit_annotation(st, returns)) {
```
Do you know why the previous code exited the annotations block and then re-entered a new one again for the return annotation?

JelleZijlstra replied:
I'm not sure. Under PEP 563 it doesn't really matter, since there are no names in annotation blocks that land in the symbol table.

carljm commented on Python/compile.c:
```diff
@@ -1600,34 +1576,78 @@ compiler_body(struct compiler *c, location loc, asdl_stmt_seq *stmts)
+        // It's possible that ste_annotations_block is set but
+        // u_deferred_annotations is not, because the former is still
+        // set if there are only non-simple annotations. However, the
+        // reverse should not be possible.
```
I don't understand this comment. What is a "non-simple" annotation?

JelleZijlstra replied:
Simple annotations are of unparenthesized bare names. Annotations for subscripts, attributes and parenthesized names are not included in `__annotations__`. This is why the AnnAssign AST node has a `simple` field.

```
Python 3.12.2 (main, May  6 2024, 20:50:52) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class X:
...     y: int
...     (z): str
... 
>>> X.__annotations__
{'y': <class 'int'>}
>>> 
```

This is explicitly discussed in PEP 526, but I'm not sure what the motivation was. Still, presumably we need to keep this behavior.

carljm commented on Python/compile.c:
```diff
@@ -2376,9 +2441,7 @@ compiler_function(struct compiler *c, stmt_ty s, int is_async)
-    if (annotations > 0) {
-        funcflags |= MAKE_FUNCTION_ANNOTATIONS;
-    }
+    funcflags |= annotations;
```
maybe rename `annotations` var to something like `annotations_flag`

carljm commented on Python/compile.c:
```diff
@@ -6468,16 +6523,35 @@ compiler_annassign(struct compiler *c, stmt_ty s)
+                }
+                else {
+                    VISIT(c, expr, s->v.AnnAssign.annotation);
+                }
```
isn't this dead code? we are already inside `if (future_annotations)` here

JelleZijlstra replied:
Right, this is a leftover from trying to implement PEP 649's REPL semantics.

JelleZijlstra committed:
CR feedback

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
Still generate __annotate__ if "from __future__ import annotations" is on

This makes it so functools.update_wrapper can always copy __annotate__ without having
to worry about whether or not the future is enabled.

JelleZijlstra committed:
Regen globals

JelleZijlstra committed:
Merge branch 'main' into pep649-compile

AlexWaygood commented on Lib/test/test_pydoc/test_pydoc.py:
```diff
@@ -76,6 +76,11 @@ class A(builtins.object)
     class B(builtins.object)
+     |  Methods defined here:
+     |
+     |  __annotate__(...)
```
(It would be great if we could generate a nice `__text_signature__` for these generated methods, but that definitely doesn't need to be done in this PR)

JelleZijlstra replied:
They're actually Python function, so I wonder why pydoc doesn't pick them up. Possibly because the parameter is called `.format`.

JelleZijlstra replied:
Indeed:

```
>>> inspect.signature(f.__annotate__)
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    inspect.signature(f.__annotate__)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
  File "/Users/jelle/py/cpython/Lib/inspect.py", line 3329, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped,
           ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                   globals=globals, locals=locals, eval_str=eval_str)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jelle/py/cpython/Lib/inspect.py", line 3055, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
                                    follow_wrapper_chains=follow_wrapped,
                                    globals=globals, locals=locals, eval_str=eval_str)
  File "/Users/jelle/py/cpython/Lib/inspect.py", line 2558, in _signature_from_callable
    return _signature_from_function(sigcls, obj,
                                    skip_bound_arg=skip_bound_arg,
                                    globals=globals, locals=locals, eval_str=eval_str)
  File "/Users/jelle/py/cpython/Lib/inspect.py", line 2403, in _signature_from_function
    parameters.append(Parameter(name, annotation=annotation,
                      ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                kind=kind))
                                ^^^^^^^^^^
  File "/Users/jelle/py/cpython/Lib/inspect.py", line 2750, in __init__
    raise ValueError('{!r} is not a valid parameter name'.format(name))
ValueError: '.format' is not a valid parameter name
```

Maybe in a followup PR we can figure out a way to make this work. The parameter has to have an illegal name so it doesn't shadow any symbol name that the user might have used.

AlexWaygood replied:
> The parameter has to have an illegal name so it doesn't shadow any symbol name that the user might have used.

I'm not sure I 100% understand this point, and no tests seem to fail if I make this change locally to your PR branch and recompile (and I verified that it means that `__annotate__` methods have valid signatures as per `inspect.signature`):

<details>

```diff
diff --git a/Include/internal/pycore_global_strings.h b/Include/internal/pycore_global_strings.h
index 009802c4416..899b19ae941 100644
--- a/Include/internal/pycore_global_strings.h
+++ b/Include/internal/pycore_global_strings.h
@@ -45,7 +45,7 @@ struct _Py_global_strings {
         STRUCT_FOR_STR(dot, ".")
         STRUCT_FOR_STR(dot_locals, ".<locals>")
         STRUCT_FOR_STR(empty, "")
-        STRUCT_FOR_STR(format, ".format")
+        STRUCT_FOR_STR(format, "format")
         STRUCT_FOR_STR(generic_base, ".generic_base")
         STRUCT_FOR_STR(json_decoder, "json.decoder")
         STRUCT_FOR_STR(kwdefaults, ".kwdefaults")
diff --git a/Include/internal/pycore_runtime_init_generated.h b/Include/internal/pycore_runtime_init_generated.h
index ff5b6ee8e0f..e8b5173a457 100644
--- a/Include/internal/pycore_runtime_init_generated.h
+++ b/Include/internal/pycore_runtime_init_generated.h
@@ -554,7 +554,7 @@ extern "C" {
     INIT_STR(dot, "."), \
     INIT_STR(dot_locals, ".<locals>"), \
     INIT_STR(empty, ""), \
-    INIT_STR(format, ".format"), \
+    INIT_STR(format, "format"), \
     INIT_STR(generic_base, ".generic_base"), \
     INIT_STR(json_decoder, "json.decoder"), \
     INIT_STR(kwdefaults, ".kwdefaults"), \
diff --git a/Python/compile.c b/Python/compile.c
index c3372766d0b..7535067ded0 100644
--- a/Python/compile.c
+++ b/Python/compile.c
@@ -1447,7 +1447,7 @@ compiler_setup_annotations_scope(struct compiler *c, location loc,
     }
     c->u->u_metadata.u_posonlyargcount = 1;
     // if .format != 1: raise NotImplementedError
-    _Py_DECLARE_STR(format, ".format");
+    _Py_DECLARE_STR(format, "format");
     ADDOP_I(c, loc, LOAD_FAST, 0);
     ADDOP_LOAD_CONST(c, loc, _PyLong_GetOne());
     ADDOP_I(c, loc, COMPARE_OP, (Py_NE << 5) | compare_masks[Py_NE]);
diff --git a/Python/symtable.c b/Python/symtable.c
index 23fc4a0ec03..0154092a905 100644
--- a/Python/symtable.c
+++ b/Python/symtable.c
@@ -2511,7 +2511,7 @@ symtable_visit_annotation(struct symtable *st, expr_ty annotation, void *key)
             }
         }
 
-        _Py_DECLARE_STR(format, ".format");
+        _Py_DECLARE_STR(format, "format");
         // The generated __annotate__ function takes a single parameter with the
         // internal name ".format".
         if (!symtable_add_def(st, &_Py_STR(format), DEF_PARAM,
@@ -2570,7 +2570,7 @@ symtable_visit_annotations(struct symtable *st, stmt_ty o, arguments_ty a, expr_
             return 0;
         }
     }
-    _Py_DECLARE_STR(format, ".format");
+    _Py_DECLARE_STR(format, "format");
     // We need to insert code that reads this "parameter" to the function.
     if (!symtable_add_def(st, &_Py_STR(format), DEF_PARAM, LOCATION(o))) {
         return 0;
```

</details>

This also might be worth a brief mention in PEP-749, since PEP-649 seems to specify that the signature of `__annotate__` should be `__annotate__(format: int)`.

In any event, this can definitely be tackled in a followup; I don't want to block this PR!

JelleZijlstra replied:
Try something like this:

```
class format: pass

def f(x: format): pass

print(f.__annotations__)
```

AlexWaygood replied:
I see, makes sense! Worth adding a test like that?

JelleZijlstra replied:
Sure, the problem is that any such test can only tell us about one specific name.

AlexWaygood replied:
Right, it doesn't verify any fundamental invariants about `__annotate__` — but I still think it's useful to have such a test, so that the CI goes obviously red if someone else comes along in the future and tries to do the "obvious fix" to get `inspect.signature()` working for these methods.

Thanks for adding it!

AlexWaygood reviewed (commented):
The bits I feel qualified to comment on all look good to me!

picnixz commented on Lib/test/test_grammar.py:
```diff
@@ -446,22 +436,12 @@ class F(C, A):
+                     {'x': int, 'y': str, 'f': typing.Tuple[int, int], 'u': int | float})
         self.assertEqual(ann_module.M.__annotations__,
-                              {'123': 123, 'o': type})
+                              {'o': type})
```
Thos indentations are a bit weird. Since you've fixed the whitespaces, maybe you could also fix them here?

JelleZijlstra replied:
Thanks, applied your suggested changes.

picnixz commented on Misc/NEWS.d/next/Core and Builtins/2024-05-22-06-22-47.gh-issue-119180.vZMiXm.rst:
```diff
@@ -0,0 +1 @@
+Evaluation of annotations is now deferred. See PEP 649 for details.
```
```suggestion
Evaluation of annotations is now deferred. See :pep:`649` for details.
```

picnixz commented on Python/compile.c:
```diff
@@ -1856,43 +1879,77 @@ compiler_visit_argannotations(struct compiler *c, asdl_arg_seq* args,
+
+    PySTEntryObject *ste;
+    int result = _PySymtable_LookupOptional(c->c_st, args, &ste);
+    if (result == -1) {
```
```suggestion
    if (_PySymtable_LookupOptional(c->c_st, args, &ste) == -1) {
```

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
Feedback from Bénédikt Tran

JelleZijlstra committed:
add test

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

carljm commented on Lib/inspect.py:
```diff
@@ -173,6 +177,13 @@
 
 
+@enum.global_enum
+class AnnotationsFormat(enum.IntEnum):
```
How does this PR relate to PEP 749? Are you intending for this PR to implement PEP 649 as written, or the updated proposals in PEP 749?

JelleZijlstra replied:
I will just take out this part; I only use the enum values in one test. The companion PR #119891 will implement PEP 749's proposed new module.

JelleZijlstra replied:
And thanks for reviewing!

carljm commented on Lib/test/test_pyrepl/test_interact.py:
```diff
@@ -105,7 +105,7 @@ def test_runsource_shows_syntax_error_for_failed_compilation(self):
     def test_no_active_future(self):
         console = InteractiveColoredConsole()
-        source = "x: int = 1; print(__annotations__)"
+        source = "x: int = 1; print(__annotate__(1))"
```
Why does printing `__annotations__` not work here? Shouldn't it call `__annotate__` under the hood anyway?

JelleZijlstra replied:
Printing a global doesn't go through the descriptor protocol, so printing `__annotations__` will just throw a NameError; it is added to the namespace only when the `__annotate__` descriptor is called.

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pep649-compile

JelleZijlstra committed:
Undo changes to inspect that we do not want

JelleZijlstra committed:
fix

JelleZijlstra committed:
Fix more tests
